contrib/sarplus/README.md

```diff
@@ -0,0 +1,191 @@
+* Fast C++ based [predictions](python/src/pysarplus.cpp)
+* Reduced memory consumption: similarity matrix cached in-memory once per worker, shared accross python executors 
+* Easy setup using [Spark Packages](https://spark-packages.org/package/Microsoft/sarplus)
+
+# Benchmarks
+
+| # Users | # Items | # Ratings | Runtime | Environment | Dataset | 
+|---------|---------|-----------|---------|-------------|---------|
+| 2.5mio  | 35k     | 100mio    | 1.3h    | Databricks, 8 workers, [Azure Standard DS3 v2](https://azure.microsoft.com/en-us/pricing/details/virtual-machines/linux/) (4 core machines) | |
+
+# Top-K Recommendation Optimization
+
+There are a couple of key optimizations:
+
+* map item ids (e.g. strings) to a continuous set of indexes to optmize storage and simplify access
+* convert similarity matrix to exactly the representation the C++ component needs, thus enabling simple shared, memory mapping of the cache file and avoid parsing. This requires a customer formatter, written in Scala
+* shared read-only memory mapping allows us to re-use the same memory from multiple python executors on the same worker node
+* partition the input test users and past seen items by users, allowing for scale out
+* perform as much of the work as possible in PySpark (way simpler)
+* top-k computation
+** reverse the join by summing reverse joining the users past seen items with any related items
+** make sure to always just keep top-k items in-memory
+** use standard join using binary search between users past seen items and the related items
+
+![Image of sarplus top-k recommendation optimization](images/sarplus_udf.svg) 
+
+# Usage
+
+```python
+import pandas as pd
+from pysarplus import SARPlus
+
+# spark dataframe with user/item/rating/optional timestamp tuples
+train_df = spark.createDataFrame(
+      pd.DataFrame({
+        'user_id': [1, 1, 2, 3, 3],
+        'item_id': [1, 2, 1, 1, 3],
+        'rating':  [1, 1, 1, 1, 1],
+    }))
+   
+# spark dataframe with user/item tuples
+test_df = spark.createDataFrame(
+      pd.DataFrame({
+        'user_id': [1, 3],
+        'item_id': [1, 3],
+        'rating':  [1, 1],
+    }))
+    
+model = SARPlus(spark, col_user='user_id', col_item='item_id', col_rating='rating', col_timestamp='timestamp')
+model.fit(train_df, similarity_type='jaccard')
+
+
+model.recommend_k_items(test_df, 'sarplus_cache', top_k=3).show()
+
+# For databricks
+# model.recommend_k_items(test_df, 'dbfs:/mnt/sarpluscache', top_k=3).show()
+```
+
+## Jupyter Notebook
+
+Insert this cell prior to the code above.
+
+```python
+import os
+
+SUBMIT_ARGS = "--packages microsoft:sarplus:0.2.6 pyspark-shell"
+os.environ["PYSPARK_SUBMIT_ARGS"] = SUBMIT_ARGS
+
+from pyspark.sql import SparkSession
+
+spark = (
+    SparkSession.builder.appName("sample")
+    .master("local[*]")
+    .config("memory", "1G")
+    .config("spark.sql.shuffle.partitions", "1")
+    .config("spark.sql.crossJoin.enabled", True)

```

Comment:
i think we should remove crossJoin.enabled config in favor of calling df.crossJoin explicitly in the code

Reply:
hmm, maybe this is not possible when using sql queries to do the join?

Reply:
I wasn't aware one can do... we can rewrite that part using pyspark functions...

Reply:
I'd love to get rid of the setting as it's kind of annoying, though I'm not sure how to do it, as crossJoin() doesn't let me specify conditions...

Reply:
let's leave it as is